compiler/rustc_const_eval/src/const_eval/eval_queries.rs

```diff
@@ -112,84 +112,56 @@ pub(super) fn op_to_const<'tcx>(
     ecx: &CompileTimeEvalContext<'_, 'tcx>,
     op: &OpTy<'tcx>,
 ) -> ConstValue<'tcx> {
+    // Handle ZST consistently and early.
+    if op.layout.is_zst() {
+        return ConstValue::ZeroSized;
+    }
+
     // We do not have value optimizations for everything.
     // Only scalars and slices, since they are very common.
-    // Note that further down we turn scalars of uninitialized bits back to `ByRef`. These can result
-    // from scalar unions that are initialized with one of their zero sized variants. We could
-    // instead allow `ConstValue::Scalar` to store `ScalarMaybeUninit`, but that would affect all
-    // the usual cases of extracting e.g. a `usize`, without there being a real use case for the
-    // `Undef` situation.
-    let try_as_immediate = match op.layout.abi {
+    let force_as_immediate = match op.layout.abi {
         Abi::Scalar(abi::Scalar::Initialized { .. }) => true,
-        Abi::ScalarPair(..) => match op.layout.ty.kind() {
-            ty::Ref(_, inner, _) => match *inner.kind() {
-                ty::Slice(elem) => elem == ecx.tcx.types.u8,
-                ty::Str => true,
-                _ => false,
-            },
-            _ => false,
-        },
```

Comment:
This is really the key change, no longer forcing a `ConstValue::Slice` for all references. That means all in-memory `OpTy` will perfectly round-trip through `ConstValue`. This used to fail in pattern matching IIRC, but that's sufficiently abstracted with valtrees now that the problem is gone. The one remaining issue was MIR printing and Clippy wanting to access the contents of an `&str`; that's what the new `try_get_slice_bytes_for_diagnostics` method is for. (Without this we would have a small regression for some clippy lints.)